### ISSUE TITLE
SSH job stuck in wait stage in dcc_collect_child

### ISSUE THREAD

**Issue by hardik-kgp** [distcc/distcc#490]
I feel that every once in a while my code gets stuck at: (I added some logging to narrow down where the code is blocked)
https://github.com/distcc/distcc/blob/e8a77c1e5fdba39a1116aa29d49c077f4196b899/src/exec.c#L479

My hunch is that the child it is waiting for gets killed before due to ssh disconnection. Hence there is no child for the calling process left. Hence the wait call is stuck. Can someone please look into this.

**Comment by hardik-kgp** [comment]
This command (sys_wait4) is blocking and doesn't respect timeout. So it keeps on waiting for child to exit. As for the ssh process, I don't think it was 'doing' anything. The compilation was complete, so ideally it should have ended all by itself. Since I already mentioned that this happens only once in a while, I am unable to reproduce this as well.


So alternate solution I am thinking to port my code is to simply use `WNOHANG` as flags, and SIGKILL the child at the end of while loop if the child is still alive.

**Comment by sourcefrog** [comment]
It looks like it's trying to call it with `WNOHANG` and periodic polling, but the code's nontrivial so maybe that is not working. Can you see what syscall it's stuck in?

**Comment by hardik-kgp** [comment]
No, that is not the case. For the case when it gets stuck, the value flags is 0, so `WNOHANG` is  not used. It was stuck in wait syscall.
